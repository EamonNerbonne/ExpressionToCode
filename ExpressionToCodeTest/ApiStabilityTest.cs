namespace ExpressionToCodeTest;

public sealed class ApiStabilityTest
{
    [Fact]
    public void PublicApi()
    {
        var publicTypes = typeof(ExpressionToCode).GetTypeInfo().Assembly.GetTypes()
            .Where(IsPublic)
            .Where(type => !type.Namespace!.Contains("Unstable"))
            .OrderByDescending(type => type.GetTypeInfo().IsEnum)
            .ThenByDescending(type => type.GetTypeInfo().IsInterface)
            .ThenBy(type => type.FullName);

        ApprovalTest.Verify(PrettyPrintTypes(publicTypes));
    }

    [Fact]
    public void UnstableApi()
    {
        var unstableTypes = typeof(ExpressionToCode).GetTypeInfo().Assembly.GetTypes()
            .Where(IsPublic)
            .Where(type => type.Namespace!.Contains("Unstable"))
            .OrderByDescending(type => type.GetTypeInfo().IsEnum)
            .ThenByDescending(type => type.GetTypeInfo().IsInterface)
            .ThenBy(type => type.FullName);

        ApprovalTest.Verify(PrettyPrintTypes(unstableTypes));
    }

    static string PrettyPrintTypes(IEnumerable<Type> types)
        => string.Join("", types.Select(PrettyPrintTypeDescription));

    static string PrettyPrintTypeDescription(Type o)
        => PrettyPrintTypeHeader(o) + "\n" + PrettyPrintTypeContents(o);

    static string PrettyPrintTypeContents(Type type)
    {
        var methods = type.GetMethods(BindingFlags.Public | BindingFlags.Instance | BindingFlags.Static)
                .OrderBy(mi => mi.MetadataToken)
                .Where(mi => mi.DeclaringType!.GetTypeInfo().Assembly != typeof(object).GetTypeInfo().Assembly) //exclude noise
            ;

        var methodBlock = string.Join("", methods.Select(mi => PrettyPrintMethod(mi) + "\n"));

        var fields = type.GetFields(BindingFlags.Public | BindingFlags.Instance | BindingFlags.Static)
                .Where(mi => mi.DeclaringType!.GetTypeInfo().Assembly != typeof(object).GetTypeInfo().Assembly) //exclude noise
            ;

        var fieldBlock = string.Join("", fields.Select(fi => PrettyPrintField(fi) + "\n"));

        return fieldBlock + methodBlock + "\n";
    }

    static string PrettyPrintTypeHeader(Type type)
    {
        var prefix = TypePrefix(type);
        var baseType = type.GetTypeInfo().BaseType;
        var inheritanceTypes = baseType == typeof(object) || baseType == null
            ? type.GetInterfaces()
            : new[] { baseType }.Concat(type.GetInterfaces().Except(baseType.GetInterfaces()));
        var suffix = !inheritanceTypes.Any() || type.GetTypeInfo().IsEnum ? "" : " : " + string.Join(", ", inheritanceTypes.Select(ObjectToCode.ToCSharpFriendlyTypeName));
        var name = type.ToCSharpFriendlyTypeName();
        return prefix + " " + name + suffix;
    }

    static string TypePrefix(Type type)
    {
        if (type.GetTypeInfo().IsEnum) {
            return "enum";
        } else if (type.GetTypeInfo().IsValueType) {
            return "struct";
        } else if (type.GetTypeInfo().IsInterface) {
            return "interface";
        } else {
            return "class";
        }
    }

    static string PrettyPrintMethod(MethodInfo mi)
    {
        var fakeTarget = mi.IsStatic ? "TYPE" : "inst";

        return "    " + mi.ReturnType.ToCSharpFriendlyTypeName() + " " + fakeTarget +
            "." + mi.Name
            + PrettyPrintGenericArguments(mi)
            + PrettyPrintParameterList(mi);
    }

    static object PrettyPrintField(FieldInfo fi)
        => "    "
            + (fi.IsLiteral ? "const " : (fi.IsStatic ? "static " : "") + (fi.IsInitOnly ? "readonly " : ""))
            + fi.FieldType.ToCSharpFriendlyTypeName()
            + " " + fi.Name
            + (fi.IsLiteral ? " = " + ObjectToCode.ComplexObjectToPseudoCode(fi.GetRawConstantValue()) : "");

    static string PrettyPrintParameterList(MethodInfo mi)
        => "(" + string.Join(
            ", ",
            mi.GetParameters()
                .Select(
                    pi =>
                        pi.ParameterType.ToCSharpFriendlyTypeName() + " " + pi.Name)) + ")";

    static string PrettyPrintGenericArguments(MethodInfo mi)
    {
        if (!mi.IsGenericMethodDefinition) {
            return "";
        }

        return "<"
            + string.Join(", ", mi.GetGenericArguments().Select(ObjectToCode.ToCSharpFriendlyTypeName))
            + ">";
    }

    static bool IsPublic(Type type)
        // ReSharper disable once ConstantNullCoalescingCondition
        => type.GetTypeInfo().IsPublic || type.GetTypeInfo().IsNestedPublic && IsPublic(type.DeclaringType ?? throw new InvalidOperationException("A nested public type has no declaring type" + type));
}
