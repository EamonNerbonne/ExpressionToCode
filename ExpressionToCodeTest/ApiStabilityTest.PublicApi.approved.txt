enum EqualityExpressionClass
    int value__
    const EqualityExpressionClass None = 0
    const EqualityExpressionClass EqualsOp = 1
    const EqualityExpressionClass NotEqualsOp = 2
    const EqualityExpressionClass ObjectEquals = 3
    const EqualityExpressionClass ObjectEqualsStatic = 4
    const EqualityExpressionClass ObjectReferenceEquals = 5
    const EqualityExpressionClass EquatableEquals = 6
    const EqualityExpressionClass SequenceEqual = 7
    const EqualityExpressionClass StructuralEquals = 8

interface IAnnotatedToCode
    string inst.AnnotatedToCode(Expression e, string msg, bool outerValueIsAssertionFailure)

interface ICodeAnnotator
    string inst.AnnotateExpressionTree(ExpressionToCodeConfiguration config, Expression expr, string msg, bool outerValueIsAssertionFailure)

interface IExpressionCompiler
    Func<T> inst.Compile<T>(Expression<Func<T>> expression)
    Delegate inst.Compile(LambdaExpression expression)

interface IExpressionToCode
    string inst.ToCode(Expression e)

interface IObjectStringifier
    string inst.PlainObjectToCode(object val, Type type)
    string inst.TypeNameToCode(Type type)
    bool inst.PreferLiteralSyntax(string str1)

class AnnotatedToCodeExtensions
    string TYPE.AnnotatedToCode(IAnnotatedToCode it, Expression e)
    string TYPE.AnnotatedToCode<T, T1, T2, T3>(IAnnotatedToCode it, Expression<Func<T, T1, T2, T3>> e)
    string TYPE.AnnotatedToCode<T, T1, T2>(IAnnotatedToCode it, Expression<Func<T, T1, T2>> e)
    string TYPE.AnnotatedToCode<T, T1>(IAnnotatedToCode it, Expression<Func<T, T1>> e)
    string TYPE.AnnotatedToCode<T>(IAnnotatedToCode it, Expression<Func<T>> e)
    string TYPE.AnnotatedToCode<T, T1, T2, T3>(IAnnotatedToCode it, Expression<Func<T, T1, T2, T3>> e, string msg, bool outerValueIsAssertionFailure)
    string TYPE.AnnotatedToCode<T, T1, T2>(IAnnotatedToCode it, Expression<Func<T, T1, T2>> e, string msg, bool outerValueIsAssertionFailure)
    string TYPE.AnnotatedToCode<T, T1>(IAnnotatedToCode it, Expression<Func<T, T1>> e, string msg, bool outerValueIsAssertionFailure)
    string TYPE.AnnotatedToCode<T>(IAnnotatedToCode it, Expression<Func<T>> e, string msg, bool outerValueIsAssertionFailure)

class CodeAnnotators
    static readonly ICodeAnnotator ValuesOnStalksCodeAnnotator
    static readonly ICodeAnnotator SubExpressionPerLineCodeAnnotator

class ExpressionExpectations
    void TYPE.Expect(Expression<Func<bool>> assertion, string msg)

class ExpressionToCode
    string TYPE.ToCode<T, T1, T2, T3>(Expression<Func<T, T1, T2, T3>> e)
    string TYPE.ToCode<T, T1, T2>(Expression<Func<T, T1, T2>> e)
    string TYPE.ToCode<T, T1>(Expression<Func<T, T1>> e)
    string TYPE.ToCode<T>(Expression<Func<T>> e)
    string TYPE.ToCode(Expression e)
    string TYPE.AnnotatedToCode<T, T1, T2, T3>(Expression<Func<T, T1, T2, T3>> e)
    string TYPE.AnnotatedToCode<T, T1, T2>(Expression<Func<T, T1, T2>> e)
    string TYPE.AnnotatedToCode<T, T1>(Expression<Func<T, T1>> e)
    string TYPE.AnnotatedToCode<T>(Expression<Func<T>> e)
    string TYPE.AnnotatedToCode(Expression expr)
    string TYPE.ToCode<T, T1, T2, T3>(ExpressionToCodeConfiguration config, Expression<Func<T, T1, T2, T3>> e)
    string TYPE.ToCode<T, T1, T2>(ExpressionToCodeConfiguration config, Expression<Func<T, T1, T2>> e)
    string TYPE.ToCode<T, T1>(ExpressionToCodeConfiguration config, Expression<Func<T, T1>> e)
    string TYPE.ToCode<T>(ExpressionToCodeConfiguration config, Expression<Func<T>> e)
    string TYPE.ToCode(ExpressionToCodeConfiguration config, Expression e)
    string TYPE.AnnotatedToCode<T, T1, T2, T3>(ExpressionToCodeConfiguration config, Expression<Func<T, T1, T2, T3>> e)
    string TYPE.AnnotatedToCode<T, T1, T2>(ExpressionToCodeConfiguration config, Expression<Func<T, T1, T2>> e)
    string TYPE.AnnotatedToCode<T, T1>(ExpressionToCodeConfiguration config, Expression<Func<T, T1>> e)
    string TYPE.AnnotatedToCode<T>(ExpressionToCodeConfiguration config, Expression<Func<T>> e)
    string TYPE.AnnotatedToCode(ExpressionToCodeConfiguration config, Expression expr)
    string TYPE.ToValuedCode<TResult>(Expression<Func<TResult>> expression)
    string TYPE.GetNameIn<TResult>(Expression<Func<TResult>> expression)
    string TYPE.GetNameIn(Expression<Action> expression)
    string TYPE.GetNameIn(Expression expr)

class ExpressionToCodeConfiguration
    static readonly ExpressionToCodeConfiguration DefaultCodeGenConfiguration
    static readonly ExpressionToCodeConfiguration DefaultAssertionConfiguration
    static ExpressionToCodeConfiguration GlobalAssertionConfiguration
    static ExpressionToCodeConfiguration GlobalCodeGenConfiguration
    ExpressionToCodeConfiguration inst.WithCompiler(IExpressionCompiler compiler)
    ExpressionToCodeConfiguration inst.WithAnnotator(ICodeAnnotator annotator)
    ExpressionToCodeConfiguration inst.WithPrintedListLengthLimit(int? limitListsToLength)
    ExpressionToCodeConfiguration inst.WithMaximumValueLength(int? limitValueStringsToLength)
    ExpressionToCodeConfiguration inst.WithOmitImplicitCasts(bool omitImplicitCasts)
    ExpressionToCodeConfiguration inst.WithObjectStringifier(IObjectStringifier objectStringifier)
    ExpressionToCodeConfiguration inst.WithAlwaysUseExplicitTypeArguments(bool alwaysUseExplicitTypeArguments)
    IExpressionToCode inst.GetExpressionToCode()
    IAnnotatedToCode inst.GetAnnotatedToCode()

class ExpressionToCodeExtensions
    string TYPE.ToCode<T, T1, T2, T3>(IExpressionToCode it, Expression<Func<T, T1, T2, T3>> e)
    string TYPE.ToCode<T, T1, T2>(IExpressionToCode it, Expression<Func<T, T1, T2>> e)
    string TYPE.ToCode<T, T1>(IExpressionToCode it, Expression<Func<T, T1>> e)
    string TYPE.ToCode<T>(IExpressionToCode it, Expression<Func<T>> e)

class ExpressionTreeAssertion
    void TYPE.Assert(ExpressionToCodeConfiguration config, Expression<Func<bool>> assertion, string msg)

class ExpressionTreeCompilers
    static readonly IExpressionCompiler DotnetExpressionCompiler
    static readonly IExpressionCompiler FastExpressionCompiler

class FormatStringParser
    (FormatStringParser.Segment[], string) TYPE.ParseFormatString(FormattableString formattableString)
    (FormatStringParser.Segment[], string) TYPE.ParseFormatString(string formatString, object[] formatArguments)

struct FormatStringParser.Segment : ValueType
    string InitialStringPart
    object FollowedByValue
    string WithFormatString

class EqualityExpressions
    EqualityExpressionClass TYPE.CheckForEquality(Expression<Func<bool>> e)
    IEnumerable<Tuple<EqualityExpressionClass, bool>> TYPE.DisagreeingEqualities(ExpressionToCodeConfiguration config, Expression<Func<bool>> e)

class ObjectStringifierExtensions
    string TYPE.PlainObjectToCode(IObjectStringifier it, object val)

class ObjectStringify
    static readonly IObjectStringifier Default
    static readonly IObjectStringifier WithFullTypeNames
    static readonly IObjectStringifier WithoutLiteralStrings
    static readonly IObjectStringifier WithFullTypeNamesWithoutLiteralStrings

class ObjectToCode
    string TYPE.ComplexObjectToPseudoCode(object val)
    string TYPE.ComplexObjectToPseudoCode(ExpressionToCodeConfiguration config, object val)
    string TYPE.PlainObjectToCode(object val)
    string TYPE.PlainObjectToCode(object val, Type type)
    string TYPE.ToCSharpFriendlyTypeName(Type type)

class PAssert
    void TYPE.IsTrue(Expression<Func<bool>> assertion)
    void TYPE.That(Expression<Func<bool>> assertion, string msg)

